plugins/qcheck-stm/src/stm_of_ir.ml

```diff
@@ -43,13 +43,294 @@ let subst inst ty =
+        (fun id ->
+          of_option
+            ~default:
+              (Ensures_not_found_for_next_state (str_of_ident id), id.id_loc)
+            (pick id))
+        value.next_state.modifies
+    in
+    let idx = List.map (fun (i, _, _) -> i) descriptions in
+    match
+      List.map (fun (_, m, e) -> (lident (str_of_ident m), e)) descriptions
+    with
+    | [] -> ok (idx, state_var)
+    | fields ->
+        (idx, pexp_record fields (Some (evar (str_of_ident state_ident)))) |> ok
+  in
+  (idx, case ~lhs ~guard:None ~rhs) |> ok
+
+let next_state config ir =
+  let cmd_name = gen_symbol ~prefix:"cmd" () in
+  let state_name = gen_symbol ~prefix:"state" () in
+  let state_ident = Gospel.Tast.Ident.create ~loc:Location.none state_name in
+  let open Reserr in
+  let* idx_cases =
+    map
+      (fun v ->
+        let* i, c = next_state_case config state_ident v in
+        ok ((v.id, i), c))
+      ir.values
+  in
+  let idx, cases = List.split idx_cases in
+  let body = pexp_match (evar cmd_name) cases in
+  let pat = pvar "next_state" in
+  let expr =
+    efun [ (Nolabel, pvar cmd_name); (Nolabel, pvar state_name) ] body
+  in
+  (idx, pstr_value Nonrecursive [ value_binding ~pat ~expr ]) |> ok
+
+let pat_char = ppat_construct (lident "Char") None
+
+let munge_longident ty lid =
+  let open Reserr in
+  match lid.txt with
+  | Lident i | Ldot (Lident i, "t") | Ldot (Ldot (_, i), "t") | Ldot (_, i) ->
+      ok (String.capitalize_ascii i)
+  | Lapply (_, _) ->
+      error
+        ( Return_type_not_supported (Fmt.str "%a" Pprintast.core_type ty),
+          ty.ptyp_loc )
+
+let mk_pat_ret_ty inst ret_ty =
+  let rec aux ty =
+    let open Reserr in
+    match ty.ptyp_desc with
+    | Ptyp_var v -> (
+        match List.assoc_opt v inst with None -> ok pat_char | Some t -> aux t)
+    | Ptyp_constr (c, xs) ->
+        let* constr_str = munge_longident ty c and* pat_xs = map aux xs in
+        let pat_arg =
+          match pat_xs with [] -> None | xs -> Some (ppat_tuple xs)
+        in
+        ppat_construct (lident constr_str) pat_arg |> ok
+    | _ ->
+        error
+          ( Return_type_not_supported (Fmt.str "%a" Pprintast.core_type ret_ty),
+            ret_ty.ptyp_loc )
+  in
+  aux ret_ty
+
+let may_raise_exception v =
+  match (v.postcond.exceptional, v.postcond.checks) with
+  | [], [] -> false
+  | _, _ -> true
+
+let postcond_case config idx state_ident new_state_ident value =
+  let idx = List.sort Int.compare idx in
+  let lhs0 = mk_cmd_pattern value in
+  let open Reserr in
+  let* lhs1 =
+    let ret_ty = Ir.get_return_type value in
+    let* ret_ty =
+      let open Ppxlib in
+      match ret_ty.ptyp_desc with
+      | Ptyp_var _ | Ptyp_constr _ -> ok ret_ty
+      | _ ->
+          error
+            ( Return_type_not_supported (Fmt.str "%a" Pprintast.core_type ret_ty),
+              ret_ty.ptyp_loc )
+    in
+    let* pat_ty = mk_pat_ret_ty value.inst ret_ty in
+    let pat_ret =
+      (* FIXME ? *)
```

Comment:
It is not obvious to me just at the moment what could be fixed there.

Reply:
I don't recall neither...